sklearn/semi_supervised/_self_training.py

```diff
@@ -0,0 +1,321 @@
+    >>> from sklearn.svm import SVC
+    >>> rng = np.random.RandomState(42)
+    >>> iris = datasets.load_iris()
+    >>> random_unlabeled_points = rng.rand(iris.target.shape[0]) < 0.3
+    >>> iris.target[random_unlabeled_points] = -1
+    >>> svc = SVC(probability=True, gamma="auto")
+    >>> self_training_model = SelfTrainingClassifier(svc)
+    >>> self_training_model.fit(iris.data, iris.target)
+    ... # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+    SelfTrainingClassifier(...)
+
+    References
+    ----------
+    David Yarowsky. 1995. Unsupervised word sense disambiguation rivaling
+    supervised methods. In Proceedings of the 33rd annual meeting on
+    Association for Computational Linguistics (ACL '95). Association for
+    Computational Linguistics, Stroudsburg, PA, USA, 189-196. DOI:
+    https://doi.org/10.3115/981658.981684
+    """
+    _estimator_type = "classifier"
+
+    def __init__(self,
+                 base_classifier,
+                 threshold=0.75,
+                 max_iter=10,
+                 n_iter_no_change=3):
+        self.base_classifier = base_classifier
+        self.threshold = threshold
+        self.max_iter = max_iter
+        self.n_iter_no_change = n_iter_no_change
+
+    def fit(self, X, y):
+        """
+        Fits this ``SelfTrainingClassifier`` to a dataset.
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
+            array representing the data
+
+        y : array-like, shape = (n_samples,)
+            array representing the labels. Unlabeled samples should have the
+            label -1.
+
+        Returns
+        -------
+        self : object
+            returns an instance of self.
+        """
+        X, y = check_X_y(X, y)
+
+        if self.base_classifier is None:
+            raise ValueError("base_classifier cannot be None!")
+
+        self.base_classifier_ = clone(self.base_classifier)
+
+        if self.max_iter is not None and self.max_iter < 0:
+            msg = "max_iter must be >= 0 or None, got {}".format(self.max_iter)
+            raise ValueError(msg)
+
+        if not 0 <= self.threshold < 1:
+            msg = "threshold must be in [0,1), got {}".format(self.threshold)
+            raise ValueError(msg)
+
+        if self.n_iter_no_change is not None and not 0 < self.n_iter_no_change:
+            msg = "n_iter_no_change must be > 0, got {}"
+            raise ValueError(msg.format(self.n_iter_no_change))
+
+        if self.max_iter is not None and self.n_iter_no_change is not None \
+                and self.n_iter_no_change >= self.max_iter:
+            msg = "n_iter_no_change >= max_iter means " \
+                "early stopping is ineffective"
+            warnings.warn(msg, UserWarning)
+
+        # XXX: y != -1 somehow doesn't work when y is an array of only strings
+        # (it returns a scalar instead of an array). Is there a better way?
+        has_label = np.vectorize(lambda x: x != -1)(y)
```

Comment:
I think `y.astype(str) != -1` is incorrect, since the API currently states that -1 (and not '-1') is the label for unlabeled samples. However, `y.astype(object) != -1` does do what i'd like. Thanks!